### PR TITLE
KAFKA-4914: Partition re-assignment tool should check types before pe…

### PR DESCRIPTION
**Pull request by nicktrav**
…rsisting state in ZooKeeper

Prior to this, there have been instances where invalid data was allowed to be persisted in ZooKeeper, which causes ClassCastExceptions when a broker is restarted and reads this type-unsafe data.

Adds basic structural and type validation for the reassignment JSON via introduction of Scala case classes that map to the expected JSON structure.

### CODE REVIEW

**Comment by nicktrav** [comment]
Prior to this, we have run into issues where invalid data was allowed to be persisted in ZooKeeper, which causes `ClassCastException`s when a broker is restarted and reads this type-unsafe data.

Adds basic structural and type validation for the reassignment JSON.

Also adds a check on `--execute` to ensure the reassignment JSON is still valid, given it can be edited after `--generate` and `--verify`.

**Comment by nicktrav** [comment]
Hey @ijuma - would you mind taking a look? I see you've worked on this previously.

**Comment by nicktrav** [comment]
@ewencp @ijuma - any chance of a review on this one?

**Comment by omkreddy** [comment]
@nicktrav Thanks for the patch. LGTM. We also faced this issue in production. Can you rebase the PR?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6169/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6153/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by nicktrav** [comment]
@omkreddy - thanks for checking this one out! I completely forgot that I had this open. I've rebased.

Would love to get it merged, as it was the source of a reasonable amount of (avoidable) pain for us a few months ago, and would love to save some others from that. Are you able to merge this in when green?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6281/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6265/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by omkreddy** [comment]
LGTM.

Though we are deprecating "--zookeeper"  option in KIP-179 (Change ReassignPartitionsCommand to use adminClient), It is good to have this check.

cc @ijuma  pinging for review.

**Comment by ijuma** [comment]
Thanks for the PR. I think we should consider using Jackson to parse into classes with the expected structure. That will do the necessary validation in a more robust way. Thoughts?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6565/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6550/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by nicktrav** [comment]
Sounds reasonable, @ijuma. I'll take another pass at it.

**Comment by nicktrav** [comment]
@ijuma - I've taken a different approach, as suggested, using case classes to hold the deserialized json. PTAL.

cc: @omkreddy

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6593/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6578/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by ijuma** [comment]
Thanks, I'll take a look tomorrow.

**Comment by nicktrav** [comment]
@ijuma - sorry to keep pinging you on this. How is this looking? Happy to address any feedback you might have.

**Comment by nicktrav** [comment]
Ping @omkreddy

**Comment by omkreddy** [comment]
@nicktrav  Thanks for your patience.  committers are busy with 0.11.0.1 release and other tasks. I am sure this will get merged into 1.0.0/October release. Pls, rebase the PR when you get a chance.


cc @ijuma

**Comment by nicktrav** [comment]
Retest this please.

**Comment by nicktrav** [comment]
@omkreddy - rebased.

**Comment by omkreddy** [comment]
cc @ijuma

File: core/src/main/scala/kafka/utils/Json.scala
Comment (ijuma) [inline core/src/main/scala/kafka/utils/Json.scala]: The idiomatic way of doing this is:

```scala
def parseTo[T](input: String)(implicit tag: ClassTag[T]): Either[Throwable, T] = {
```

Then the caller can just invoke `parseTo[PartitionAssignment](jsonData)`.

Also, should we call it `parseAs` perhaps?

File: core/src/main/scala/kafka/utils/ZkUtils.scala
Comment (ijuma) [inline core/src/main/scala/kafka/utils/ZkUtils.scala]: You should use pattern matching to retrieve `left` or `right` so that you can avoid the unsafe `get` method below.

File: core/src/main/scala/kafka/utils/ZkUtils.scala
Comment (ijuma) [inline core/src/main/scala/kafka/utils/ZkUtils.scala]: I think we should not document these classes as being for deserialization only. We should also use them for serialization so that we detect problems at that stage too.

Also, we can move these classes to `ZkData`.

File: core/src/test/scala/unit/kafka/utils/JsonTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/utils/JsonTest.scala]: It's better to place this inside a companion object so that it's restricted to this test.

File: core/src/test/scala/unit/kafka/utils/JsonTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/utils/JsonTest.scala]: Maybe we can also have a test where we can roundtrip (serialize and then deserialize)? It would be nice to have a method in `Json` to serialize with `Jackson`. Maybe we can even actually replace the existing one.

File: core/src/test/scala/unit/kafka/utils/ZkUtilsTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/utils/ZkUtilsTest.scala]: Nit: we don't use `_` in method names.

**Review by ijuma — COMMENTED** [review]
Sorry for the long delay. Finally got around to reviewing this. Left a few comments.

**Comment by nicktrav** [comment]
Thanks for the review @ijuma! I've addressed your comments in separate commits.

For the case of your JSON parsing comment, I've added a new dependency on jackson-module-scala that deals with Scala collections a little better, rather than the existing custom parsing logic. Not sure how you feel about adding a new dependency, although there are already a couple of other Jackson libraries in use already.

**Comment by ijuma** [comment]
Let's remove the last commit for now as there is an ongoing discussion on whether we should use jackson-scala (https://github.com/apache/kafka/pull/4259). The rest looks good.

**Comment by nicktrav** [comment]
Done! Thanks for the review and getting this one over the line, @ijuma. 👍

**Comment by ijuma** [comment]
@nicktrav, I had a closer look and it looks like this PR doesn't achieve the original goal as the Controller no longer uses `ZkUtils`. I think we should revert the `ZkUtils` changes and do something like https://github.com/nicktrav/kafka/pull/1

I didn't run the tests, but it gives you an idea of what I had in mind. Thoughts?

**Comment by nicktrav** [comment]
@ijuma - sounds good! I didn't realize it had changed. Did you want to re-use the commit you pushed? Otherwise, I might add a few more unit tests to your changes and open a new PR.

**Comment by ijuma** [comment]
You can just use whatever is useful in your PR. Thanks!

**Comment by ijuma** [comment]
@nicktrav I just merged https://github.com/apache/kafka/commit/0a508a436c928c0440e8b90dda98c22dc0ec244c which converts to Jackson for serialization as well. Now would be a good time to submit your PR which would be the first one to transition from Maps to actual classes.

**Comment by nicktrav** [comment]
@ijuma - awesome. Thanks for the follow-up. I'll try and get to this sometime this week.

**Comment by nicktrav** [comment]
@ijuma - incorporated your changes and fixed up the remaining merge conflicts. This should be good for a final round of review. Thanks for persevering!

**Comment by nicktrav** [comment]
Hey @ijuma - Happy New Year. Gentle ping on this one again. I've rebased this a few times now, so if you could take another look, I'd appreciate it. Thanks!

File: core/src/main/scala/kafka/utils/Json.scala
Comment (ijuma) [inline core/src/main/scala/kafka/utils/Json.scala]: The result type can be `Either[JsonProcessingException, T]`.

File: core/src/main/scala/kafka/zk/ZkData.scala
Comment (ijuma) [inline core/src/main/scala/kafka/zk/ZkData.scala]: Field name should be `partition`

File: core/src/main/scala/kafka/utils/ZkUtils.scala
Comment (ijuma) [inline core/src/main/scala/kafka/utils/ZkUtils.scala]: Seems like we'd want to include the original exception as the cause. Also, why is it a `ConfigException`?

File: core/src/main/scala/kafka/zk/ZkData.scala
Comment (ijuma) [inline core/src/main/scala/kafka/zk/ZkData.scala]: We should be using this in `ReassignPartitionsZNode` as in my PR.

File: core/src/main/scala/kafka/utils/ZkUtils.scala
Comment (ijuma) [inline core/src/main/scala/kafka/utils/ZkUtils.scala]: We should use `ReassignPartitionsZNode.decode`

File: core/src/test/scala/unit/kafka/utils/ZkUtilsTest.scala
Comment (ijuma) [inline core/src/test/scala/unit/kafka/utils/ZkUtilsTest.scala]: We should rewrite these tests to use `ReassignPartitionsZNode`. Also, since they don't require `ZooKeeper`, they should be unit tests. Perhaps `ReassignPartitionsZNodeTest`.

**Review by ijuma — COMMENTED** [review]
Sorry for the delay @nicktrav. I left a few comments.

**Comment by nicktrav** [comment]
Thanks for the review @ijuma. I pushed a new commit addressing your last round of comments. I'm sure some of it is not idiomatic Scala, so would appreciate some guidance, if any.

**Comment by omkreddy** [comment]
Can we use zkUtils.parsePartitionReassignmentData here
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala#L245
above change will solve KAFKA-6084. This can be done in another PR.

**Comment by ijuma** [comment]
@omkreddy Ideally, we should not be using `ZkUtils` in `ReassignPartitionsCommand` (i.e. we should use `KafkaZkClient`). Also, `ReassignPartitionsCommand` also parses a `log_dirs` field which is not handled in `ZkUtils` (or `KafkaZkClient`). Let's do that clean-up in a separate PR.

**Comment by nicktrav** [comment]
@omkreddy @ijuma - any action for me to take here based on the last two comments?

**Comment by omkreddy** [comment]
LGTM. +1

**Comment by ijuma** [comment]
Sorry, I have just moved countries and won't have time until next week.

**Comment by nicktrav** [comment]
No worries at all. Good luck with the move!

**Comment by nicktrav** [comment]
Any chance of an update on this @ijuma @omkreddy? (coincidentally, today is the 1 year birthday of the ticket! 🍰)

**Comment by omkreddy** [comment]
maybe @hachikuji  can help to merge this patch.

File: core/src/main/scala/kafka/utils/Json.scala
Comment (viktorsomogyi) [inline core/src/main/scala/kafka/utils/Json.scala]: Hmm, just a very basic question: why are you using here Either/Left/Right instead of Try/Success/Failure?
As from my previous experiences we used Try when we wanted to propagate the exception.
And please not I don't mind using Either, just asking for reasons. :)

**Review by viktorsomogyi — APPROVED** [review]
Although (as mentioned on my JIRA) this PR seems to aim bit different goals then mine, I reviewed this PR and made one minor comment/question.
All in all I think the PR looks good. Would be happy to see this merged.

File: core/src/main/scala/kafka/utils/Json.scala
Comment (nicktrav) [reply core/src/main/scala/kafka/utils/Json.scala]: My (idiomatic) Scala isn't great, but this came at the suggestion of @ijuma (above):
https://github.com/apache/kafka/pull/2708#discussion_r151342320

Comment (ijuma) [reply core/src/main/scala/kafka/utils/Json.scala]: `Try` doesn't let you specify an exception subclass and behaves a bit different if an exception is thrown during `map`, etc.

**Review by ijuma — APPROVED** [review]
I pushed minor changes and it now LGTM. Will merge it after the tests pass. Thanks so much for your patience.

**Comment by nicktrav** [comment]
Thanks @ijuma!

**Pull request merged**
